src/main/java/org/terracotta/statistics/derived/histogram/BarSplittingBiasedHistogram.java

```diff
@@ -0,0 +1,252 @@
+  private final double alphaPhi;
+  private final double ratio;
+  private final List<Bar> bars;
+  private final double[] maxSizeTable;
+  
+  private long size;
+  
+  public BarSplittingBiasedHistogram(double maxCoefficient, double phi, int expansionFactor, int bucketCount, float barEpsilon, long window) {
+    this.bucketCount = bucketCount;
+    this.barCount = bucketCount * expansionFactor;
+    
+    this.bars = new ArrayList<Bar>(barCount);
+    this.bars.add(new Bar(barEpsilon, window));
+    this.phi = phi;
+    this.alphaPhi = (phi == 1.0f) ? 1.0 / bucketCount : (1 - phi) / (1 - Math.pow(phi, bucketCount));
+
+    double rho = Math.pow(phi, 1.0 / expansionFactor);
+    double alphaRho = (rho == 1.0f) ? 1.0 / barCount : (1 - rho) / (1 - Math.pow(rho, barCount));
+    this.ratio = (rho / (1.0 + rho));
+    this.maxSizeTable = new double[barCount];
+    for (int i = 0; i < barCount; i++) {
+      this.maxSizeTable[i] = maxCoefficient * alphaRho * Math.pow(rho, i);
+    }
+  }
+
+  public BarSplittingBiasedHistogram(int bucketCount, long window) {
+    this(1.7f, 0.7f, 7, bucketCount, 0.1f, window);
+  }
+
+  public BarSplittingBiasedHistogram(double phi, int bucketCount, long window) {
+    this(1.7f, phi, 7, bucketCount, 0.1f, window);
+  }
+  
+  public void event(long value, long time) {
+    int barIndex = getBarIndex(value);
+    Bar bar = bars.get(barIndex);
+    long before = bar.count();
+    bar.insert(value, time);
+    long after = bar.count();
+    size += (after - before);
+    if (after > maxBarSize(barIndex)) {
+      split(bar, barIndex);
+    }
+  }
+
+  @Override
+  public String toString() {
+    return bars.toString();
+  }
+  
+  @Override
+  public List<Histogram.Bucket<Double>> getBuckets() {
+    List<Histogram.Bucket<Double>> buckets = new ArrayList<Histogram.Bucket<Double>>(bucketCount);
+    double targetSize = size() * alphaPhi; // * phi^0
+    Iterator<Bar> it = bars.iterator();
+    Bar b = it.next();
+    double minimum = b.minimum();
+    double count = b.count();
+    for (int i = 0; i < bucketCount - 1; i++) {
+      while (count < targetSize && it.hasNext()) {
+        count += (b = it.next()).count();
+      }
+      
+      double surplus = count - targetSize;
+      double maximum = b.minimum() + (b.maximum() - b.minimum()) * (1 - surplus/b.count());
+      buckets.add(new ImmutableDoubleBucket(minimum, maximum, targetSize));
+      minimum = maximum;
+      count = surplus;
+      targetSize *= phi;
+    }
+    while (it.hasNext()) {
+      count += (b = it.next()).count();
+    }
+    buckets.add(new ImmutableDoubleBucket(minimum, b.maximum(), count));
+    return buckets;
+  }
+
+  public double[] getQuantileBounds(double quantile) {
+    double threshold = quantile * size();
+    long count = 0;
+    for (Bar b : bars) {
+      count += b.count();
+      if (count >= threshold) {
+        return new double[] {b.minimum(), b.maximum()};
+      }
+    }
+    throw new AssertionError();
```

Comment:
I think it would be better because I guess this method call is public at the end and will be used to configure which percentile we need.